### PR TITLE
feat: bring ZkTrie in

### PR DESCRIPTION
**Pull request by omerfirmak**
## 1. Purpose or design rationale of this PR

This PR moves in the ZkTrie implementation from its standalone repo and implements the `Commit()`function to behave as the Geth expects it to behave.

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [ ] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [ ] Yes

## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [ ] This PR is not a breaking change
- [ ] Yes

## Release Notes

- **New Features**
	- Introduced a new `Byte32` type for managing 32-byte data structures, including hashing and conversion methods.
	- Enhanced `ZkTrie` functionality with improved error handling and state management.
	- Added methods for managing nodes in a Merkle Trie, including hashing and node type handling.

- **Bug Fixes**
	- Updated error handling across various trie operations to provide more specific feedback.
	- Refactored trie database interactions to eliminate deprecated structures and improve performance.

- **Tests**
	- Comprehensive test coverage added for new functionalities, including unit tests for `Byte32`, trie nodes, and `ZkTrie` operations.
	- Restructured existing tests to ensure clarity and maintainability, focusing on edge cases and error conditions.

- **Chores**
	- Removed deprecated code and files related to the old `ZkTrie` implementation.
	- Updated dependency management in `go.mod` to reflect current library usage.

### CODE REVIEW

**Comment by georgehao** [comment]
@coderabbitai review

**Comment by georgehao** [comment]
@coderabbitai review

**Comment by georgehao** [comment]
@coderabbitai review

File: internal/ethapi/api_test.go
Comment (0xmountaintop) [inline internal/ethapi/api_test.go:1]: `TestTransaction_RoundTripRpcJSON` fails but it doesn't fail on `syncUpstream/active`. If we should use `EmptyRootHash` here maybe there're some other places to fix?

File: rollup/tracing/proof.go [hunk 1]
```diff
@@ -153,17 +152,11 @@ func (t *ProofTracer) Prove(key []byte, proofDb ethdb.KeyValueWriter) error {
...
 			mptPath = append(mptPath, n)
 			t.rawPaths[string(key)] = mptPath
 		},
 	)
-	if err != nil {
-		return err
-	}
-	// we put this special kv pair in db so we can distinguish the type and
-	// make suitable Proof
-	return proofDb.Put(magicHash, zktrie.ProofMagicBytes())
```
Comment (colinlyguo) [inline rollup/tracing/proof.go:168]: Would this bring incompatible changes? or are they just outdated codes?

File: core/state_processor.go [hunk 2]
```diff
@@ -144,7 +144,7 @@ func applyTransaction(msg *Message, config *params.ChainConfig, gp *GasPool, sta
 	// Apply the transaction to the current state (included in the env).
 	applyMessageStartTime := time.Now()
 	result, err := ApplyMessage(evm, msg, gp, l1DataFee)
-	if evm.Config.Tracer != nil && evm.Config.Tracer.IsDebug() {
+	if evm.Config.Tracer != nil {
```
Comment (colinlyguo) [inline core/state_processor.go:147]: Will this open a counter-based ccc tracer for all nodes?

File: rollup/tracing/proof.go [hunk 3]
```diff
@@ -103,46 +101,47 @@ func (t *ProofTracer) GetDeletionProofs() ([][]byte, error) {
 }
 
 // MarkDeletion mark a key has been involved into deletion
-func (t *ProofTracer) MarkDeletion(key []byte) {
+func (t *ProofTracer) MarkDeletion(key []byte) error {
```
Comment (colinlyguo) [inline rollup/tracing/proof.go:104]: does adding an error return for interface compatibility? because I see in the function, it only returns nil when success and panic in error paths.

**Review by colinlyguo — DISMISSED** [review]

File: internal/ethapi/api_test.go
Comment (omerfirmak) [reply internal/ethapi/api_test.go:1]: fixed

File: trie/sync_test.go [hunk 4]
```diff
@@ -57,7 +60,7 @@ func makeTestTrie(scheme string) (ethdb.Database, *Database, *StateTrie, map[str
 		}
 	}
 	root, nodes, _ := trie.Commit(false)
-	if err := triedb.Update(root, types.EmptyLegacyTrieRootHash, 0, trienode.NewWithNodeSet(nodes), nil); err != nil {
+	if err := triedb.Update(root, types.EmptyStateRootHash, 0, trienode.NewWithNodeSet(nodes), nil); err != nil {
```
Comment (0xmountaintop) [inline trie/sync_test.go:63]: `EmptyStateRootHash` doesn't exist

File: trie/trie_reader.go [hunk 5]
```diff
@@ -45,14 +46,10 @@ type trieReader struct {
...
-	// 	if stateRoot == (common.Hash{}) {
-	// 		log.Error("Zero state root hash!")
-	// 	}
-	// 	return &trieReader{owner: owner}, nil
-	// }
-	if stateRoot == types.EmptyZkTrieRootHash {
-		// log.Error("Zero state root hash!")
+	if stateRoot == (common.Hash{}) || stateRoot == types.EmptyRootHash {
+		if stateRoot == (common.Hash{}) {
+			log.Error("Zero state root hash!")
```
Comment (0xmountaintop) [inline trie/trie_reader.go:51]: isn't zktrie empty root hash `common.Hash{}`? why do we need to log error here?

File: trie/zk_trie.go [hunk 6]
```diff
@@ -1,210 +1,1044 @@
...
+// GetKey returns the key for the given hash.
+func (mt *ZkTrie) GetKey(hashKey []byte) []byte {
+	mt.lock.RLock()
+	defer mt.lock.RUnlock()
+	return mt.getKey(hashKey)
 }
 
-func (t *ZkTrie) UpdateAccount(address common.Address, acc *types.StateAccount) error {
-	return t.TryUpdateAccount(address.Bytes(), acc)
+// GetKey returns the key for the given hash.
```
Comment (colinlyguo) [inline trie/zk_trie.go:554]:
```suggestion
// getKey returns the key for the given hash.
```

File: core/state/statedb_test.go [hunk 7]
```diff
@@ -1051,7 +1051,7 @@ func TestFlushOrderDataLoss(t *testing.T) {
 		t.Fatalf("failed to commit state trie: %v", err)
 	}
 	triedb.Reference(root, common.Hash{})
-	if err := triedb.Cap(1024); err != nil {
+	if err := triedb.Cap(128); err != nil {
```
Comment (0xmountaintop) [inline core/state/statedb_test.go:1054]: what's the reasone behind lowering this limit?

**Review by 0xmountaintop — DISMISSED** [review]

**Review by colinlyguo — DISMISSED** [review]

**Comment by colinlyguo** [comment]
`@github-actions CI / build-mock-ccc-geth (pull_request) Failing after 26s`
this CI failure is worth fixing.

**Review by georgehao — DISMISSED** [review]

File: trie/zk_trie.go [hunk 8]
```diff
@@ -1,210 +1,1044 @@
...
+func NewZkTrie(id *ID, db *Database) (*ZkTrie, error) {
+	reader, err := newTrieReader(id.StateRoot, id.Owner, db)
+	if err != nil {
+		return nil, err
+	}
+
+	mt := ZkTrie{
+		owner:        id.Owner,
+		reader:       reader,
+		maxLevels:    NodeKeyValidBytes * 8,
```
Comment (0xmountaintop) [inline trie/zk_trie.go:98]: how is this decided?

**Review by colinlyguo — APPROVED** [review]

**Comment by colinlyguo** [comment]
> `@github-actions CI / build-mock-ccc-geth (pull_request) Failing after 26s` this CI failure is worth fixing.

da-codec breaks it. fixed.

File: trie/trie_reader.go (hunk 5 above)
Comment (omerfirmak) [reply trie/trie_reader.go:51]: We don't, we can remove it if it ends up being too annoying.

File: rollup/tracing/proof.go (hunk 3 above)
Comment (omerfirmak) [reply rollup/tracing/proof.go:104]: yeah, looks unneeded. none of the callsites check the returned error either. let's keep it for now, it is rather harmless.

File: core/state_processor.go (hunk 2 above)
Comment (omerfirmak) [reply core/state_processor.go:147]: It was already enabled on all nodes if I am not mistaken. This change doesn't really affect that.

File: core/state/statedb_test.go (hunk 7 above)
Comment (omerfirmak) [reply core/state/statedb_test.go:1054]: zkTrie nodes are smaller, so we need a lower limit here to trigger a flush to the disk.

File: rollup/tracing/proof.go (hunk 1 above)
Comment (omerfirmak) [reply rollup/tracing/proof.go:168]: Yeap, outdated AFAICS.

**Review by 0xmountaintop — APPROVED** [review]
